scripts/airgapped/get-all-images.py

```diff
@@ -0,0 +1,177 @@
+import argparse
+import logging
+import shutil
+import subprocess
+
+import git
+import yaml
+
+# logging
+LOG_FORMAT = "%(levelname)s \t| %(message)s"
+logging.basicConfig(format=LOG_FORMAT, level=logging.INFO)
+log = logging.getLogger(__name__)
+
+# consts
+EXCLUDE_REPOS = ["kubeflow-roles"]
+
+GH_REPO_KEY = "_github_repo_name"
+GH_BRANCH_KEY = "_github_repo_branch"
+GH_DEPENDENCY_REPO_KEY = "_github_dependency_repo_name"
+GH_DEPENDENCY_BRANCH_KEY = "_github_dependency_repo_branch"
+GET_IMAGES_SH = "tools/get-images.sh"
+
+AIRGAP_TESTING_IMAGES = [
+    "charmedkubeflow/pipelines-runner:ckf-1.8",
+    "docker.io/kubeflowkatib/simple-pbt:v0.16.0",
+    "ghcr.io/knative/helloworld-go:latest",
+    "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0",
+]
+
+
+def is_dependency_app(app: dict) -> bool:
+    """Detect if app in bundle is not owned by Analytics team."""
+    if GH_DEPENDENCY_REPO_KEY in app and GH_DEPENDENCY_BRANCH_KEY in app:
+        return True
+
+    return False
+
+
+def bundle_app_contains_gh_metadata(app: dict) -> bool:
+    """
+    Given an application in a bundle check if it contains github metadata keys
+    to be able to be parsed properly.
+    """
+    if is_dependency_app(app):
+        return True
+
+    if GH_REPO_KEY in app and GH_BRANCH_KEY in app:
+        return True
+
+    return False
+
+
+def clone_repo(repo_name: str, branch: str) -> str:
+    """Clones locally a repo and returns the folder created."""
+    repo_url = f"https://github.com/canonical/{repo_name}.git"
+
+    logging.info(f"Cloning repo {repo_url}")
+    repo = git.Repo.clone_from(repo_url, str(repo_name))
+
+    logging.info(f"Checking out to branch {branch}")
+    repo.git.checkout(branch)
+
+    return repo_name
+
+
+def get_analytics_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a charm developed by us, by:
+    1. Cloning the repo of the charm
+    2. Running the repo's tools/get-images.sh
+    3. Delete the repo
+
+    If the tools/get-images.sh of a repo fails for any reason then this
+    script will also fail.
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_REPO_KEY], app[GH_BRANCH_KEY])
+
+    logging.info(f"Executing repo's {GET_IMAGES_SH} script")
+    process = subprocess.Popen(["bash", "tools/get-images.sh"], cwd=repo_dir,
+                               stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = process.communicate()
+    if stderr:
+        raise ValueError("Script '%s' for charm '%s' had error logs: \n%s" %
```

Comment:
I think that there are two approaches at this. If we want to use `subprocess`, maybe you want to use `run` instead of `Popen`. [`run`](https://docs.python.org/3/library/subprocess.html#subprocess.run) is a much more simple implementation that runs the subprocess and waits for it to finish, which I think it's what we want here.

It also raises [`CalledProcessError`](https://docs.python.org/3/library/subprocess.html#subprocess.CalledProcessError), so it helps a lot with exception handling when the subprocess returns a non-zero exit code (which we expect the `get-images.sh` will do if something fails). This is translated to:

```python
try:
  subprocess.run(command)
except subprocess.CalledProcessError:
  # exception handling here
```

The other option is to use a python package called [`sh`](https://sh.readthedocs.io/en/latest/#sh), which also has [exceptions](https://sh.readthedocs.io/en/latest/#exit-codes) for handling non-zero exit codes. The advantage of `sh` is that its syntax is much more friendly for people that are familiar with the CLI.

Reply:
Ah nice! I used `subprocess.run` with the small caveat that I wanted to also print the `stderr` if the process failed, so that we can easily see what failed from this script (and not have to manually run the command that failed).

Because of that I had to:
1. Catch the `CalledProcessError`
2. Log a message that includes the `stderr` of the process
3. Re-raise the caught exception

25d5863